CIAB: Use postgres as the default traffic vault backend

srijeet0406 opened the pull request:
<!--
************ STOP!! ************
If this Pull Request is intended to fix a security vulnerability, DO NOT submit it! Instead, contact
the Apache Software Foundation Security Team at security@trafficcontrol.apache.org and follow the
guidelines at https://www.apache.org/security/ regarding vulnerability disclosure.
-->
## What does this PR (Pull Request) do?
<!-- Explain the changes you made here. If this fixes an Issue, identify it by
replacing the text in the checkbox item with the Issue number e.g.

- [x] This PR fixes #9001 OR is not related to any Issue

^ This will automatically close Issue number 9001 when the Pull Request is
merged (The '#' is important).

Be sure you check the box properly, see the "The following criteria are ALL
met by this PR" section for details.
-->

- [x] This PR is not related to any Issue <!-- You can check for an issue here: https://github.com/apache/trafficcontrol/issues -->


## Which Traffic Control components are affected by this PR?
<!-- Please delete all components from this list that are NOT affected by this
Pull Request. Also, feel free to add the name of a tool or script that is
affected but not on the list.

Additionally, if this Pull Request does NOT affect documentation, please
explain why documentation is not required. -->

- CDN in a Box

## What is the best way to verify this PR?
<!-- Please include here ALL the steps necessary to test your Pull Request. If
it includes tests (and most should), outline here the steps needed to run the
tests. If not, lay out the manual testing procedure and please explain why
tests are unnecessary for this Pull Request. -->

Spin up CIAB and make sure all the containers are up and running without errors.
Now, exec into the `cdn-in-a-box_db` container and make sure you see all the tables in the `traffic_vault` database. 
Create a delivery service from TP in CIAB and generate SSL keys for it. Now make sure that those keys exist in the `sslkey` table in the `traffic_vault` database. Do similar tests for URI sig keys and DNSSEC keys.
 
## If this is a bug fix, what versions of Traffic Control are affected?
<!-- If this PR fixes a bug, please list here all of the affected versions - to
the best of your knowledge. It's also pretty helpful to include a commit hash
of where 'master' is at the time this PR is opened (if it affects master),
because what 'master' means will change over time. For example, if this PR
fixes a bug that's present in master (at commit hash '1df853c8'), in v4.0.0,
and in the current 4.0.1 Release candidate (e.g. RC1), then this list would
look like:

- master (1df853c8)
- 4.0.0
- 4.0.1 (RC1)

If you don't know what other versions might have this bug, AND don't know how
to find the commit hash of 'master', then feel free to leave this section
blank (or, preferably, delete it entirely).
 -->
- master

## The following criteria are ALL met by this PR
<!-- Check the boxes to signify that the associated statement is true. To
"check a box", replace the space inside of the square brackets with an 'x'.
e.g.

- [ x] <- Wrong
- [x ] <- Wrong
- [] <- Wrong
- [*] <- Wrong
- [x] <- Correct!

-->

- [x] This PR includes tests 
- [x] This PR does not include documentation
- [x] This PR does not include an update to CHANGELOG.md
- [x] This PR includes any and all required license headers
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)


## Additional Information
<!-- If you would like to include any additional information on the PR for
potential reviewers please put it here.

Some examples of this would be:

- Before and after screenshots/gifs of the Traffic Portal if it is affected
- Links to other dependent Pull Requests
- References to relevant context (e.g. new/updates to dependent libraries,
mailing list records, blueprints)

Feel free to leave this section blank (or, preferably, delete it entirely).
-->

<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

zrhoffman commented on traffic_ops/install/bin/_postinstall:
```diff
@@ -1004,6 +1018,17 @@ def generate_cdn_conf(questions, fname, automatic, root): # type: (list[Question
+		exception.__cause__ = e
+		logging.warn("no traffic vault backend configured, using default postgres")
+
+	if traffic_vault_backend == "postgres":
+		return True
+	else:
+		return False
```
This can be `return traffic_vault_backend == "postgres"`

zrhoffman commented on infrastructure/cdn-in-a-box/traffic_ops/run-go.sh:
```diff
@@ -138,11 +143,13 @@ cd /opt/traffic_ops/app;
 
-traffic_ops_golang_command=(./bin/traffic_ops_golang -cfg "$CDNCONF" -dbcfg "$DBCONF" -riakcfg "$RIAKCONF");
+traffic_ops_golang_command=(./bin/traffic_ops_golang -cfg "$CDNCONF" -dbcfg "$DBCONF");
+#-riakcfg "$RIAKCONF");
```
Can this commented line be removed?

zrhoffman commented on infrastructure/cdn-in-a-box/traffic_ops/config.sh:
```diff
@@ -76,6 +81,19 @@ cdn_conf=/opt/traffic_ops/app/conf/cdn.conf
+            "user": "$TV_DB_USER",
+            "password": "$TV_DB_USER_PASS",
+            "port": ${TV_DB_PORT:-5432},
+            "ssl": false,
```
```json
"ssl": false,
```
is already in `cdn.conf`, no need to specify here

zrhoffman commented on infrastructure/cdn-in-a-box/traffic_ops/config.sh:
```diff
@@ -76,6 +81,19 @@ cdn_conf=/opt/traffic_ops/app/conf/cdn.conf
+            "password": "$TV_DB_USER_PASS",
+            "port": ${TV_DB_PORT:-5432},
+            "ssl": false,
+            "conn_max_lifetime_seconds": 60,
```
`"conn_max_lifetime_seconds"` should be set to `${DEBUGGING_TIMEOUT:-60}` so debugging sessions don't time out.

zrhoffman commented on infrastructure/cdn-in-a-box/traffic_ops/config.sh:
```diff
@@ -76,6 +81,19 @@ cdn_conf=/opt/traffic_ops/app/conf/cdn.conf
+            "conn_max_lifetime_seconds": 60,
+            "max_connections": 500,
+            "max_idle_connections": 30,
+            "query_timeout_seconds": 10
```
`"query_timeout_seconds"` should be set to `${DEBUGGING_TIMEOUT:-60}` so debugging sessions don't time out.

zrhoffman commented on infrastructure/cdn-in-a-box/traffic_ops/config.sh:
```diff
@@ -32,9 +32,14 @@
 # TO_HOST
 # TO_PORT
 # TP_HOST
+# TV_DB_USER
+# TV_DB_NAME
+# TV_DB_SERVER
+# TV_DB_USER_PASS
+# TV_DB_PORT
```
Can these be alphabetized? Some of the rest of the list is also unsorted, might as well sort it all.

zrhoffman commented on infrastructure/cdn-in-a-box/traffic_ops/run-go.sh:
```diff
@@ -30,16 +30,21 @@
 # ADMIN_USER
 # ADMIN_PASS
 # DOMAIN
+# TV_DB_USER
+# TV_DB_NAME
+# TV_BACKEND
+# TV_DB_SERVER
+# TV_DB_USER_PASS
+# TV_DB_PORT
```
Same here, environment variables should be sorted.

zrhoffman commented on traffic_ops/install/bin/_postinstall:
```diff
@@ -1004,6 +1018,17 @@ def generate_cdn_conf(questions, fname, automatic, root): # type: (list[Question
 	logging.info("CDN configuration has been saved")
+	try:
+		traffic_vault_backend = existing_conf["traffic_ops_golang"]["traffic_vault_backend"]
+	except KeyError as e:
+		exception = ValueError("missing 'traffic_vault_backend' config_var")
+		exception.__cause__ = e
+		logging.warn("no traffic vault backend configured, using default postgres")
```
The `exception` built here should be raised, the warning is not sufficient.

zrhoffman reviewed (changes requested):
`traffic_ops_golang` is not running in the `trafficops` container. From the logs (downloadable as a GHA artifact):

```go
trafficops_1      | 2021-05-17T20:33:47.807659531Z ERROR: traffic_ops_golang.go:286: 2021-05-17T20:33:46.928632269Z: failed to get Traffic Vault backend 'postgres': failed to load backend 'postgres': validating Postgres config: 'aes_key_location' cannot be blank
```

srijeet0406 commented on traffic_ops/install/bin/_postinstall:
```diff
@@ -1004,6 +1018,14 @@ def generate_cdn_conf(questions, fname, automatic, root): # type: (list[Question
 		json.dump(existing_conf, conf_file, indent=indent)
 		print(file=conf_file)
 	logging.info("CDN configuration has been saved")
+	try:
```
That was my thinking behind leaving it as a warning.

srijeet0406 replied:
I have set it back to a warning

zrhoffman replied:
> but since `traffic_vault_backend` is not a required field, should we actually be raising an exception in this case?

Good point, warning makes sense.

srijeet0406 commented on traffic_ops/app/conf/cdn.conf:
```diff
@@ -10,6 +10,19 @@
         "workers" : 12
     },
     "traffic_ops_golang" : {
+        "traffic_vault_backend": "postgres",
```
Nope, didn't want to commit that. removed now.

srijeet0406 commented:
> `traffic_ops_golang` is not running in the `trafficops` container. From the logs (downloadable as a GHA artifact):
> 
> ```go
> trafficops_1      | 2021-05-17T20:33:47.807659531Z ERROR: traffic_ops_golang.go:286: 2021-05-17T20:33:46.928632269Z: failed to get Traffic Vault backend 'postgres': failed to load backend 'postgres': validating Postgres config: 'aes_key_location' cannot be blank
> ```

Yeah I think I'll need to add the config changes for Matt's AES PR.

zrhoffman commented on infrastructure/cdn-in-a-box/traffic_ops/aes.conf:
```diff
@@ -0,0 +1 @@
+CZhbsfrdNqAC9tfZJ/x4tB/NgoXhbfLdFwauhhNedjE=
```
CDN in a Box should generate this. If you want, it can be persisted in a bind mount like the CA directory is:

https://github.com/apache/trafficcontrol/blob/c3878a2a997d48e106ea44bd72e34aa39598e2b9/infrastructure/cdn-in-a-box/docker-compose.yml#L73-L75

But if it does not exist, CiaB should generate it, and no generated key should be tracked in the repo.

zrhoffman commented on infrastructure/cdn-in-a-box/traffic_ops/Dockerfile:
```diff
@@ -120,6 +120,9 @@ COPY infrastructure/cdn-in-a-box/enroller/server_template.json \
 	infrastructure/cdn-in-a-box/traffic_ops/tv.conf \
 	/
 
+# Generate the AES encryption key for traffic vault backend
+RUN echo $(openssl rand -base64 32) > aes.conf
```
We should avoid storing any secrets in the built Docker image. This should be run from a script at runtime.

rawlinp reviewed (commented):
`infrastructure/cdn-in-a-box/traffic_ops/aes.conf` is an empty file -- did you mean to commit that?

Also, `.conf` makes it seem like a config file, but in reality it is a key file, so it might make more sense to use the `.key` suffix.

srijeet0406 commented:
> `infrastructure/cdn-in-a-box/traffic_ops/aes.conf` is an empty file -- did you mean to commit that?
> 
> Also, `.conf` makes it seem like a config file, but in reality it is a key file, so it might make more sense to use the `.key` suffix.

I keep doing that, fixed now. Also, changed the suffix of the file.

rawlinp commented on traffic_ops/app/conf/cdn.conf:
```diff
@@ -10,6 +10,20 @@
         "workers" : 12
     },
     "traffic_ops_golang" : {
+        "traffic_vault_backend": "",
```
you might've miscommitted these changes again

srijeet0406 replied:
I think I might've misunderstood your previous comment. I thought you were referring to the values being committed, and not the keys. I get it now, will remove.

ocket8888 commented on traffic_ops/install/bin/_postinstall:
```diff
@@ -710,6 +721,30 @@ def unmarshal_config(dct): # type: (dict) -> dict[str, list[Question]]
+	logging.info(aes_key_location)
+
+	cmd = ["/usr/bin/openssl", "rand", "-base64", "32"]
+	key = open(aes_key_location, 'w')
```
File descriptor leak happens here if any of the following lines throw exceptions e.g. if `/usr/bin/openssl` does not exist or is not executable. This should either use context management like:
```python3
with open(aes_key_location, "w") as key:
    # do stuff
```
(not sure if ~~the dead version of Python nobody should ever use~~ Python2 supports that) or a `try`...`finally` like:
```python3
key = open(aes_key_location, "w")
try:
    # do stuff
finally:
    key.close()
```

rawlinp replied:
Python2 supports `with`

ocket8888 commented on traffic_ops/install/bin/_postinstall:
```diff
@@ -905,7 +940,7 @@ def random_word(length = 12): # type: (int) -> str
+def generate_cdn_conf(questions, fname, automatic, root): # type: (list[Question], str, bool, str) -> bool
 	"""
 	Generates some properties of a cdn.conf file based on the passed questions.
 
```
This doc string should specify what the function `:returns:`, now that it does.

rawlinp commented on traffic_ops/app/conf/cdn.conf:
```diff
@@ -34,8 +34,6 @@
         },
         "whitelisted_oauth_urls": [],
         "oauth_client_secret": "",
-        "traffic_vault_backend": "",
```
I think we may want to keep these in here just to indicate that this is where traffic_vault is configured, but to keep them empty indicating that it's optional.

rawlinp commented on infrastructure/cdn-in-a-box/traffic_ops/run-go.sh:
```diff
@@ -142,7 +148,12 @@ mkdir -p /var/log/traffic_ops
 
-traffic_ops_golang_command=(./bin/traffic_ops_golang -cfg "$CDNCONF" -dbcfg "$DBCONF" -riakcfg "$RIAKCONF");
+if [[ -z $TV_BACKEND ]]; then
+  traffic_ops_golang_command=(./bin/traffic_ops_golang -cfg "$CDNCONF" -dbcfg "$DBCONF" -riakcfg "$RIAKCONF");
```
so if both `-riakcfg` and `traffic_vault_backend` (in cdn.conf) are used, I believe `traffic_vault_backend` will take precedence. So we would also have to exclude the `traffic_vault_backend`-related config in cdn.conf in order to use this line to enable Riak. Or, we can actually specify the riak config in `traffic_vault_backend` in cdn.conf (and omit the `riakcfg`) if we would like to use Riak instead of postgres.

rawlinp replied:
Nevermind, we chatted about this offline, and now I understand why this works the way it is. 😄

zrhoffman commented on traffic_ops/install/bin/_postinstall:
```diff
@@ -710,6 +721,30 @@ def unmarshal_config(dct): # type: (dict) -> dict[str, list[Question]]
+	"""
+	logging.info(aes_key_location)
+
+	cmd = ["/usr/bin/openssl", "rand", "-base64", "32"]
```
Can this use the existing `exec_openssl()` function?

zrhoffman commented on infrastructure/cdn-in-a-box/traffic_ops/tv.conf:
```diff
@@ -0,0 +1,11 @@
+
```
Nit: First line of the file is empty

ocket8888 commented on traffic_ops/app/conf/cdn.conf:
```diff
@@ -93,4 +93,4 @@
         }
     ]
-}
+}
```
Missing newline at EOF

rawlinp commented on traffic_ops/install/bin/_postinstall:
```diff
@@ -710,6 +721,27 @@ def unmarshal_config(dct): # type: (dict) -> dict[str, list[Question]]
+	"""
+	logging.info(aes_key_location)
+
+	cmd = ["/usr/bin/openssl", "rand", "-base64", "32"]
```
this `cmd` is no longer used

rawlinp commented on traffic_ops/install/bin/_postinstall:
```diff
@@ -1004,6 +1039,19 @@ def generate_cdn_conf(questions, fname, automatic, root): # type: (list[Question
+	try:
+		traffic_vault_backend = existing_conf["traffic_ops_golang"]["traffic_vault_backend"]
+	except KeyError as e:
+		logging.warn("no traffic vault backend configured, using default postgres")
```
I think this is supposed to be `logging.warning`

rawlinp commented:
I'm getting this error when trying to build the trafficops container in CiaB:
```
Step 23/24 : COPY tv.conf conf/production/
COPY failed: stat /var/lib/docker/tmp/docker-builder422447319/tv.conf: no such file or directory
ERROR: Service 'trafficops' failed to build : Build failed
```

srijeet0406 commented:
> I'm getting this error when trying to build the trafficops container in CiaB:
> 
> ```
> Step 23/24 : COPY tv.conf conf/production/
> COPY failed: stat /var/lib/docker/tmp/docker-builder422447319/tv.conf: no such file or directory
> ERROR: Service 'trafficops' failed to build : Build failed
> ```

Thats weird. I did just a clean build and everything works fine for me. 
```
docker-compose down --remove-orphans
docker volume prune
docker system prune
make very-clean
make all && docker-compose build --parallel && docker-compose -f docker-compose.yml -f docker-compose.expose-ports.yml up -d
```

srijeet0406 commented:
I feel like the CI checks would have freaked out too, had the container building failed.

zrhoffman commented:
> > I'm getting this error when trying to build the trafficops container in CiaB:
> > ```
> > Step 23/24 : COPY tv.conf conf/production/
> > COPY failed: stat /var/lib/docker/tmp/docker-builder422447319/tv.conf: no such file or directory
> > ERROR: Service 'trafficops' failed to build : Build failed
> > ```
> 
> Thats weird. I did just a clean build and everything works fine for me.
> 
> ```
> docker-compose down --remove-orphans
> docker volume prune
> docker system prune
> make very-clean
> make all && docker-compose build --parallel && docker-compose -f docker-compose.yml -f docker-compose.expose-ports.yml up -d
> ```

Sounds like @rawlinp has `DOCKER_BUILDKIT=1` and `COMPOSE_DOCKER_CLI_BUILD=1` in his environment and you do not. Set those environment variables to reproduce and add the file as an exception to `infrastructure/cdn-in-a-box/traffic_ops/Dockerfile.dockerignore` to fix

rawlinp commented:
I don't have those env vars set, but when I change the line to
```
COPY infrastructure/cdn-in-a-box/traffic_ops/tv.conf conf/production/
```
it builds. I'm currently working through what appeared to be expired certs in my CiaB which was causing the startup to get stuck in an infinite loop.

srijeet0406 commented:
I just set those variables and rebuilt and it still builds and works fine for me.

zrhoffman commented:
Ahh sounds like I was off, glad you have it sorted out.

srijeet0406 commented:
> I don't have those env vars set, but when I change the line to
> 
> ```
> COPY infrastructure/cdn-in-a-box/traffic_ops/tv.conf conf/production/
> ```
> 
> it builds. I'm currently working through what appeared to be expired certs in my CiaB which was causing the startup to get stuck in an infinite loop.

I thought I took care of that in the upper `COPY` block:

```
COPY infrastructure/cdn-in-a-box/enroller/server_template.json \
	infrastructure/cdn-in-a-box/traffic_ops/config.sh \
	infrastructure/cdn-in-a-box/traffic_ops/run-go.sh \
	infrastructure/cdn-in-a-box/traffic_ops/to-access.sh \
	infrastructure/cdn-in-a-box/dns/insert-self-into-dns.sh \
	infrastructure/cdn-in-a-box/dns/set-dns.sh \
	infrastructure/cdn-in-a-box/traffic_ops/set-to-ips-from-dns.sh \
	infrastructure/cdn-in-a-box/traffic_ops/generate-certs.sh \
	infrastructure/cdn-in-a-box/traffic_ops/adduser.pl \
	infrastructure/cdn-in-a-box/traffic_ops/trafficops-init.sh \
	infrastructure/cdn-in-a-box/variables.env \
	infrastructure/cdn-in-a-box/traffic_ops/tv.conf \
	/
```

rawlinp commented:
That copies it to `/`, but then later in the Dockerfile it's trying to copy to `conf/production/`. Would it make sense to just copy to `conf/production/` to start? I have no idea why my build isn't working otherwise.

srijeet0406 commented:
> That copies it to `/`, but then later in the Dockerfile it's trying to copy to `conf/production/`. Would it make sense to just copy to `conf/production/` to start? I have no idea why my build isn't working otherwise.
Right, hence I added the second copy, because I wanted to group all the "base" copies together. But yes, I will copy it directly to `conf/production`.
